Data Structures/Trees/LCA/BinaryLifting.cpp

```diff
@@ -0,0 +1,183 @@
+// This program finds LCA of two nodes in a rooted tree using Binary Lifting
+
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+const int L = 1e5+7;
+std::vector<int> v[L];
+int parent[L][22], level[L];
+void dfs(int vertex, int prev_node)
+{
+	// assigning depth of current node wrt root
+	level[vertex] = level[prev_node] + 1;
+
+	// immediate parent in the dfs tree is the 2^0th = 1st ancestor 
+	parent[vertex][0] = prev_node;
+	for(auto &x : v[vertex])
+	{
+		if(x != prev_node)
+		{
+			dfs(x, vertex);
+		}
+	}
+}
+int findLCA(int v1, int v2)
+{
+	if(level[v1] > level[v2])
+	{
+		swap(v1, v2);
+	}
+	// if both nodes are not at same level then bring them at same level 
+	int dist = level[v2] - level[v1];
+	for (int i = 0; i < 22; ++i)
+	{
+		if(dist & 1)
+		{
+			v2 = parent[v2][i]; 
+		}
+		dist >>= 1;
+	}
+	// if v2's ancestor on the same level as v1 is v1 it means that v1 lies in the
+	// path from v2 to the root and hence the LCA of v1 and v2 is v1
+	if( v1 == v2 )return v1;
+
+	for (int i = 21; i >= 0; --i)
+	{
+		// Jump higher only if the 2^i th ancestor of both the nodes are not the same. This is done because 
+		// the nodes in the path from the LCA to the root are all common ancestors of the 2 nodes and if 
+		// we need to reach the LCA we should avoid jumping to a node at a higher level than the LCA. 
+		// This condition helps us avoid it. Ultimately we reach a node which is the 1st child of 
+		// LCA and then it's easy to get the LCA as it would be the 2^0 th ancestor of the node. 
+		if(parent[v1][i] != parent[v2][i])
```

Comment:
You can also mention that when we reach away from the root of the tree, i.e. `parent[v1][i] = -1`, jump doesn't take place and we get the correct result.

Reply:
I have added the condition @sahilbansal17

Reply:
Good work. 👍

Reply:
By the way, the condition would be handled **implicitly**, what do you say? 
As both *v<sub>1</sub>* and *v<sub>2</sub>* will be pointing to **-1** and hence they will be equal.

Reply:
That's why I didn't add that condition.But this would give a clearer explanation I guess

Reply:
Yes, its better now.